### DISCUSSION TITLE
Classification on new unseen dataset

### DISCUSSION THREAD

**Discussion by usyed15181** [pycaret/pycaret#1221] [Q&A] [answered]
I have trained a model with Normalization, Transformation, Remove multicollinearity and Fix imbalance at setup.

exp_clf = setup(df, target = 'Identity'
               , normalize=True
               , fix_imbalance=True
               , transformation = True
               , remove_multicollinearity=True)

Created the model:
lr = create_model('lr')

Now when i bring in new unseen dataset, how can i process new data with same setup config before applying the trained model?

predictions on new dataset:
lr_pred_new = predict_model(lr, data = new_data)  #new_data is pd dataframe

The problem is new_data also needs Normalization, Transformation, Remove multicollinearity and Fix imbalance. How can i do this?

Thanks

**Comment by Yard1** [comment, marked as answer]
All of those transformations will be applied when you use `predict_model`.

**Reply by usyed15181** [reply]
Thanks, thats awesome. 

All i need to do is to only select same features in new_data as in trained data? or thats already taken care of too?

**Reply by Yard1** [reply]
You only need to pass the unseen dataframe in the same format as the dataframe used in setup (so same column names, types, etc.) and everything will be taken care of.

**Reply by usyed15181** [reply]
Thanks alot. Another question.

Now i want to evaluate model performance on new unseen data? I want to see Accuracy, AUC, Recall, Prec., F1, Kappa, MCC for new dataset like its shown with create_model()?

Should i use evaluate_model() ? Giving me errors though.

**Reply by Yard1** [reply]
`predict_model` will give you those scores, displaying a dataframe with them. You can obtain that dataframe by doing `pull()` after `predict_model`.

**Reply by usyed15181** [reply]
Amazing support! Thanks for quick help. Its working for me!

**Reply by suresh-chinta** [reply]
Thanks for the thread and answers. I have a similar question on predict_model. When I run the below code.
I get the same exact accuracy metrics, which is not correct because one is a validation sample and another is an unseen test sample.

![image](https://user-images.githubusercontent.com/19486358/117314576-868a1d00-aea4-11eb-88e5-22fad96fc7b1.png)

Can you please help with this?

**Reply by Yard1** [reply]
In the second case you are getting the same metrics because predicting on unseen data doesn't display scores, so if you do pull() it simply displays the scores for the first predict_model. Please check the documentation, there is a method in the utils module that will allow you to score the predictions obtained through predict_model.

**Reply by usyed15181** [reply]
Suresh. I saved the model first then loaded it back to run on unseen data and run pull. Try that.

**Reply by suresh-chinta** [reply]
Thank you both.

**Reply by usyed15181** [reply]
Did it work Suresh?

**Reply by suresh-chinta** [reply]
Thanks for checking. I tried the below. saving and loading did not automatically print the metrics table for unseen test-set. Also, the AUC obtained using this approach on hold-out set is not matching with the AUC reported using predict_model on hold-out set.

save_model(lr, 'lr_5fold.pkl')
lr = load_model('lr_5fold.pkl')
oot_predit_lr = predict_model(lr, data = account_df_test)
roc_auc_score(oot_predit_lr['Label'], oot_predit_lr['Score'])

**Reply by usyed15181** [reply]
it wont match. Your hold out set is different than your account_df_test. Isn't it?

**Reply by usyed15181** [reply]
and to print metrics, you have to do:
metrics = pull()
metrics.head()

**Reply by suresh-chinta** [reply]
hi, I mean to say, for neither hold-out set predictions nor unseen test set predictions the AUC reported by Pycaret, did match with the AUC computed using roc_auc_score(true_y, predicted_y). I guess it has to do with the threshold.

hold_out_preds = predict_model(lr)
roc_auc_score(hold_out_preds['Label'], hold_out_preds['Score'])

**Reply by pit9921** [reply]
you can get your metric's as follows: the example is for regression, but could be easily set up to classification metrics:

# generate predictions on unseen data
predictions = predict_model(final_best, data = unseen)
from sklearn.metrics import mean_absolute_error
mean_absolute_error(predictions['Durchmesser'],predictions['Label']) 
from sklearn.metrics import r2_score
r2_score(predictions['Durchmesser'],predictions['Label'])
from sklearn.metrics import mean_squared_error 
mean_squared_error(predictions['Durchmesser'],predictions['Label']) 
# RSME
mean_squared_error(predictions['Durchmesser'],predictions['Label'], squared = False)

**Reply by rahulsahai78** [reply]
Hello, 
I am not able to find a way to see the metrics for the unseen data.
Is there any function to pull the metrics?
thank you,

**Reply by moezali1** [reply]
If you have true labels of unseen data then you can call check_metric function from pycaret.utils module or just use sklearn.metrics directly.
